docs/rfcs/20220209_limit_connections_rate.md

```diff
@@ -0,0 +1,91 @@
+- Feature Name: Ability to limit connection creation rate per shard
+- Status: in-progress
+- Start Date: 2022-02-09
+- Authors: Vadim Plakhtinskiy
+- Issue: 
+
+# Executive Summary
+
+- Set a limit on the rate with which the shard accepts new connections
+- Set a limit on connection creation rate per IP
+
+
+Similar to KIP-612, adapted to Redpanda's thread per core model
+
+## What is being proposed
+
+New configuration properties (For start we choose global config for connection rate):
+- kafka_max_connection_creation_rate (optional integer, default unset)
+- kafka_max_connection_creation_rate_ovverides (string, default "")
+
+
+When a connection is accepted which exceeds one of the configured
```

Comment:
For rate limiting, I think we probably want to block rather than dropping -- if someone is only slightly exceeding the limit then this is friendlier and produces better fairness.  Then have a limit on how many can be waiting in this state, and if _that_ limit is exceeded then drop.

(I went with dropping for the connection count limiting RFC because hitting the connection count limit is an indefinite "no", whereas hitting a rate limit is only a temporary "no")

Reply:
semaphore has a .waiters() method, so you could use that to decide whether to block or drop.

Reply:
Ok, I did not know that we want block connection. Looks like, `wait(duration timeout, size_t nr=1) noexcept` for semaphore can help